connectivity/tests/lrp.go

```diff
@@ -0,0 +1,190 @@
+		frontend := check.NewLRPFrontend(spec.RedirectFrontend)
+		frontendStr := net.JoinHostPort(frontend.Address(features.IPFamilyV4), fmt.Sprint(frontend.Port()))
+		lrpBackendsMap := make(map[string][]string)
+		// Check for LRP backend pods deployed on nodes in the cluster.
+		for _, pod := range t.Context().LrpBackendPods() {
+			node := pod.NodeName()
+			podIP := pod.Pod.Status.PodIP
+			if _, ok := lrpBackendsMap[node]; !ok {
+				lrpBackendsMap[node] = []string{podIP}
+			}
+			lrpBackendsMap[node] = append(lrpBackendsMap[node], podIP)
+		}
+		// Wait until the local redirect entries are plumbed in the BPF LB map
+		// on the cilium agent nodes hosting LRP backend pods.
+		WaitForLocalRedirectBPFEntries(ctx, t, frontendStr, lrpBackendsMap)
+	}
+
+	// Tests client pods to LRP frontend connectivity: traffic gets redirected
+	// to the LRP backends.
+	for _, pod := range t.Context().LrpClientPods() {
+		pod := pod
+
+		for _, policy := range policies {
+			policy := policy
+
+			if policy.Spec.SkipRedirectFromBackend != s.skipRedirectFromBackend {
+				continue
+			}
+
+			i := 0
+			lf := check.NewLRPFrontend(policy.Spec.RedirectFrontend)
+			t.NewAction(s, fmt.Sprintf("curl-%d", i), &pod, lf, features.IPFamilyV4).Run(func(a *check.Action) {
+				a.ExecInPod(ctx, ct.CurlCommand(lf, features.IPFamilyV4))
+				i++
+			})
+		}
+	}
+
+	// Tests LRP backend pods to LRP frontend connectivity: traffic gets redirected
+	// based on the configured skipRedirectFromBackend flag.
+	for _, pod := range t.Context().LrpBackendPods() {
+		pod := pod
+
+		for _, policy := range policies {
+			policy := policy
+
+			if policy.Spec.SkipRedirectFromBackend != s.skipRedirectFromBackend {
+				continue
+			}
+
+			i := 0
+			lf := check.NewLRPFrontend(policy.Spec.RedirectFrontend)
+			t.NewAction(s, fmt.Sprintf("curl-%d", i), &pod, lf, features.IPFamilyV4).Run(func(a *check.Action) {
+				a.ExecInPod(ctx, ct.CurlCommand(lf, features.IPFamilyV4))
+
+				if policy.Spec.SkipRedirectFromBackend {
+					a.ValidateFlows(ctx, pod, a.GetEgressRequirements(check.FlowParameters{
+						AltDstIP:   lf.Address(features.IPFamilyV4),
+						AltDstPort: lf.Port(),
+					}))
+				}
+				i++
+			})
+		}
+
+	}
+}
+
+func WaitForLocalRedirectBPFEntries(ctx context.Context, t *check.Test, frontend string, backendsMap map[string][]string) {
+	ct := t.Context()
+	w := wait.NewObserver(ctx, wait.Parameters{Timeout: 20 * time.Second})
+	defer w.Cancel()
+
+	ensureBPFLBEntries := func() error {
+		for _, ciliumPod := range ct.CiliumPods() {
+			node := ciliumPod.Pod.Spec.NodeName
+			backends, ok := backendsMap[node]
+			if !ok {
+				// No LRP backend pods deployed on this node.
+				continue
+			}
+			cmd := strings.Split("cilium bpf lb list -o json", " ")
+			stdout, err := ciliumPod.K8sClient.ExecInPod(ctx, ciliumPod.Pod.Namespace, ciliumPod.Pod.Name, defaults.AgentContainerName, cmd)
+			if err != nil {
+				t.Fatal("Failed to run cilium bpf lb list -o json command:", err)
+			}
```

Comment:
This pattern is not ideal as it is relies on finicky timeouts and mechanisms to parse output in an error-prone way which we have seen cause many flakes in the Ginkgo suite over time. Ofttimes with timeouts, it's inevitable that we will have a PR to bump it because the initial value was too low. Then we get to a very bloated duration which leads to tests taking longer and longer over time. With parsing command-line outputs, it's inevitable that we get flakes where the command doesn't execute properly, i.e. exiting with a strange error code, empty output, etc, that we presumed are ultimately underlying infrastructure problems. 

Now we are porting this sort of logic over to the Cilium CLI, which goes against the goals of the Cilium CLI that's meant to be a less error-prone style of testing than the Ginkgo suite.

On a more practical note, it seems that this logic is trying to detect when the LRP entries have been created. Are there alternative mechanisms to detecting when LRP is "ready" besides dumping the BPF map? For example, with CNPs we can detect the policy revision number has incremented. If an alternative doesn't exist today, it might be worth discussing what it would look like and temporarily moving ahead with this PR as-is (pending other feedback of course).

Reply:
>For example, with CNPs we can detect the policy revision number has incremented.

Is that done in userspace or datapath (BPF maps)? Does that correspond to individual policy entries? We don't have such equivalent metric on LB side of things. 

That's a fair question. 

Some more context on the datapath side of things in general - 

Checking userspace state is not enough, as it can take some time to translate plumb policies into datapath. You wouldn't want to have a busy loop checking if a test curl command eventually succeeded (or fails for that matter) within a timeout. So the more definite signals are checking for BPF entries (LB, egress gateway, etc tests follow similar logic). Another useful aspect with these kind of checks is that they are the next logical step in debugging when connectivity tests fail. For example, if I'm debugging an LB connectivity issue, I would go about checking BPF LB map entries, and whether they were plumbed correctly.

Reply:
>In simpler terms, if you change/apply a policy, all endpoints wait until they have been regenerated for the policy change (i.e. wait for the revision to increment). Once they have all finished regenerating, the policy revision exposed via the API is incremented which is what the CLI watches.

Thanks for the context. I don't see how the same logic could translate to other BPF maps (e.g., LB). Does that mean we would have to serialize service LB plumbing? The point around checking BPF entries as a logical step for debugging LB still stands though. 

>Ofttimes with timeouts, it's inevitable that we will have a PR to bump it because the initial value was too low. 

Going back to your concerns around timeouts, won't you have similar timeouts while watching policy revision bumps?

Reply:
> Going back to your concerns around timeouts, won't you have similar timeouts while watching policy revision bumps?

Technically yes, but the waiting mechanism is built into `cilium-dbg policy wait ...`, so there's a lot less testing code that needs to be written which means a lot less surface area for potential problems/errors.

Reply:
> So the more definite signals are checking for BPF entries (LB, egress gateway, etc tests follow similar logic)

It sounds like some sort of first-class 'ready' signal would benefit LB, LRP, egress gateway features all the same. Anyway, it's something to think about and for the immediate PR, it doesn't sound like we have other alternatives so we are left with the current implementation for now.

Reply:
> > Going back to your concerns around timeouts, won't you have similar timeouts while watching policy revision bumps?
> 
> Technically yes, but the waiting mechanism is built into `cilium-dbg policy wait ...`, so there's a lot less testing code that needs to be written which means a lot less surface area for potential problems/errors.

Ah, TIL!! Thanks for highlighting this.

Reply:
@aditighag I remembered that I didn't ask for a tracking issue regarding this, but it would be nice to track the work to migrating this sort of testing style to something less error-prone, so we don't forget about it.

Reply:
I've filed - https://github.com/cilium/cilium/issues/33372.